src/csharp/Grpc.Tools.MSBuildIntegration/Grpc.Tools.MSBuildIntegration.targets

```diff
@@ -0,0 +1,27 @@
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <CoreCompileDependsOn>Proto;$(CoreCompileDependsOn)</CoreCompileDependsOn>
+  </PropertyGroup>
+  <ItemGroup>
+    <AvailableItemName Include="Proto"/>
+  </ItemGroup>
+  <Target Name="Proto" Inputs="@(Proto)" Outputs="@(Proto->'$(IntermediateOutputPath)%(FileName).cs')">
+    <!-- Run our platform independent wrapper that launches protoc with the right arguments -->
+    <Exec Condition="'$(OS)' == 'Windows_NT'" Command="$(MSBuildThisFileDirectory)..\tools\Grpc.Tools.CSharpGenerator.exe $(ProjectDir) $(IntermediateOutputPath) @(Proto->'%(filename)%(extension)', ' ')" LogStandardErrorAsError="true"/>
+    
+    <!-- Make sure that our protoc and protoc_csharp_plugin files are executable -->
+    <Exec Condition="'$(OS)' == 'Unix'" Command="chmod +x $(MSBuildThisFileDirectory)../tools/*/protoc"/>
+    <Exec Condition="'$(OS)' == 'Unix'" Command="chmod +x $(MSBuildThisFileDirectory)../tools/*/grpc_csharp_plugin"/>    
+
+    <!-- Run our platform independent wrapper that launches protoc with the right arguments -->
+    <Exec Condition="'$(OS)' == 'Unix'" Command="mono $(MSBuildThisFileDirectory)../tools/Grpc.Tools.CSharpGenerator.exe $(ProjectDir) $(IntermediateOutputPath) @(Proto->'%(filename)%(extension)', ' ')" LogStandardErrorAsError="true"/>
+    
+    <!-- Ensure that only files that exist (are touchable) are added in to the Compile target -->
+    <Touch Files="@(Proto->'$(IntermediateOutputPath)%(FileName).cs')" ForceTouch="true">

```

Comment:
I have an idea. Using `<Touch>` to test only if the file exists is a little bit of a hack. I think a more straightforward way to add a file to the collection is, *inside a `<Target>`*,
<ItemGroup>
  <Compile Condition="Exists('$(IntermediateOutputPath)%(ProtoFileName).cs')" Include="$(IntermediateOutputPath)%(ProtoFileName).cs" />
</ItemGroup>
Not really sure if this is VS2010 compatible, if that version still exists anywhere on Earth. If you decided to go with the <Touch>, please add `AlwaysCreate="false"`, as [the spec](https://msdn.microsoft.com/en-us/library/37fwbyt5.aspx) does not state the default value for this attribute, and you do not want to create an empty file here by mistake.